test/e2e/app-dir/resume-data-cache/resume-data-cache.test.ts

```diff
@@ -0,0 +1,398 @@
+import { nextTestSetup } from 'e2e-utils'
+import { measurePPRTimings } from 'e2e-utils/ppr'
+import { parsePostponedState } from 'next/dist/server/app-render/postponed-state'
+
+/**
+ * Comprehensive test suite for Next.js resume data cache functionality.
+ *
+ * This test suite covers:
+ * - Cache creation and serialization during static generation
+ * - Cache restoration during dynamic rendering
+ * - Integration with fetch caching and "use cache" directive
+ * - PPR streaming with cache resume
+ * - Error handling and edge cases
+ * - Production-specific scenarios
+ *
+ * Note: Uses both "use cache" directive (modern) and unstable_cache (legacy)
+ * depending on the specific features being tested (tags, revalidation, etc.)
+ */
+describe('resume-data-cache', () => {
+  const { next, isNextDev, isNextStart } = nextTestSetup({
+    files: __dirname,
+    skipDeployment: true,
+  })
+
+  // Skip tests in dev mode as PPR is not available
+  if (isNextDev) {
+    it('should skip tests in dev mode', () => {})
+    return
+  }
+
+  describe('Cache Creation and Storage', () => {
+    it('should create and populate prerender cache during static generation', async () => {
+      const $ = await next.render$('/static-with-cache')
+
+      // Check that static content is rendered
+      expect($('#static-content').text()).toBe('Static Content')
+
+      // Check that cached fetch data is included
+      expect($('#cached-fetch').text()).toContain('Cached at:')
+
+      // Verify the cache was used by checking the metadata
+      try {
+        const meta = await next.readFile(
+          '.next/server/app/static-with-cache.meta'
+        )
+        const metadata = JSON.parse(meta)
+
+        // Should have postponed state with resume data
+        expect(metadata.postponed).toBeDefined()
+        expect(metadata.postponed).toContain('store')
+      } catch (e) {
+        // Meta files might not exist in all configurations
+        console.log('Meta file not found, skipping metadata check')
+      }
+    })
+
+    it('should serialize and compress cache data correctly', async () => {
+      try {
+        const meta = await next.readFile(
+          '.next/server/app/static-with-cache.meta'
+        )
+        const metadata = JSON.parse(meta)
+
+        // Use the official utility to parse the postponed state
+        const postponedData = metadata.postponed
+        if (postponedData && typeof postponedData === 'string') {
+          const parsedState = parsePostponedState(postponedData, undefined)
+          const parsedCache = parsedState.renderResumeDataCache
+
+          // Verify cache structure using the parsed cache
+          expect(parsedCache.cache).toBeDefined()
+          expect(parsedCache.fetch).toBeDefined()
+          expect(parsedCache.encryptedBoundArgs).toBeDefined()
+          expect(parsedCache.decryptedBoundArgs).toBeDefined()
```

Comment:
Sequential assertions are not helpful in the failure case. Imagine `.cache` is undefined all of the sudden but `decryptedBoundArgs` is not. You wouldn't see it.

Jest has a better failure message if you group them as much as possible e.g. `expect(parsedState).toEqual(expect.objectContaining({ renderResumeDataCache: expect.objectContaining({ cache: ???, fetch: ???, encryptedBoundArgs: ???, decryptedBoundArgs: ??? }) }))`

For the individual properties you could either use strict equality or some more loose matchers like `expect.any(String)`.